Support integral type range window (#2020)
NVIDIA/spark-rapids: docs/compatibility.md

between time zones is not currently supported on the GPU. Therefore operations involving timestamps
will only be GPU-accelerated if the time zone used by the JVM is UTC.

## Windowing

### Window Functions

Because of ordering differences between the CPU and the GPU window functions especially row based
window functions like `row_number`, `lead`, and `lag` can produce different results if the ordering
includes both `-0.0` and `0.0`, or if the ordering is ambiguous. Spark can produce different results
from one run to another if the ordering is ambiguous on a window function too.

### Range Window

> When the order-by column of a range based window is numeric type like `byte/short/int/long` and the range boundary calculated for a value has overflow, CPU and GPU will get different results.

Reviewer: Please wrap lines to approx. 100 characters to match the style elsewhere in this document.  It makes it easier to read this file from a terminal and also easier to comment on lines for review.

Author: Done.

> For example, let's assume we have two columns, one partition-by column named `uid` and the other order-by column named `dollars`. And the dataset is as below.

Reviewer: IMO we don't need to elaborate on partition ID and order-by, that's already covered in the SQL statement below.
```suggestion
For example, consider the following dataset:
```

Author: Done

``` console
+------+---------+
| id   | dollars |
+------+---------+
|    1 |    NULL |
|    1 |      13 |
|    1 |      14 |
|    1 |      15 |
|    1 |      15 |
|    1 |      17 |
|    1 |      18 |
|    1 |      52 |
|    1 |      53 |
|    1 |      61 |
|    1 |      65 |
|    1 |      72 |
|    1 |      73 |
|    1 |      75 |
|    1 |      78 |
|    1 |      84 |
|    1 |      85 |
|    1 |      86 |
|    1 |      92 |
|    1 |      98 |
+------+---------+
```

> After executing below SQL statement,

Reviewer: ```suggestion
After executing the SQL statement:
```

Author: Done

``` sql
SELECT
 COUNT(dollars) over
    (PARTITION BY id
    ORDER BY CAST (dollars AS Byte) ASC
    RANGE BETWEEN 127 PRECEDING AND 127 FOLLOWING)

> From table

Reviewer: ```suggestion
FROM table
```

Author: Done

```

> We can get below different results for both CPU and GPU.

Reviewer: ```suggestion
The results will differ between the CPU and GPU due to overflow handling.
```

Author: Done

``` console
CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0])
GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
```

> 

Reviewer: This section should mention the configs added in this PR and/or link to the configs document so users reading this are aware that they can force the CPU behavior if desired.

Author: Done

the range boundary calculated for a value has overflow, CPU and GPU will get different results.

For example, consider the following dataset:

``` console
+------+---------+
| id   | dollars |
+------+---------+
|    1 |    NULL |
|    1 |      13 |
|    1 |      14 |
|    1 |      15 |
|    1 |      15 |
|    1 |      17 |
|    1 |      18 |
|    1 |      52 |
|    1 |      53 |
|    1 |      61 |
|    1 |      65 |
|    1 |      72 |
|    1 |      73 |
|    1 |      75 |
|    1 |      78 |
|    1 |      84 |
|    1 |      85 |
|    1 |      86 |
|    1 |      92 |
|    1 |      98 |
+------+---------+
```

After executing the SQL statement:

``` sql
SELECT
 COUNT(dollars) over
    (PARTITION BY id
    ORDER BY CAST (dollars AS Byte) ASC
    RANGE BETWEEN 127 PRECEDING AND 127 FOLLOWING)
FROM table
```

The results will differ between the CPU and GPU due to overflow handling.

``` console
CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0])
GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
```

> To disable this operation on the GPU, set

Reviewer: "this operation" is a little too generic, best to be explicit here with something like, "To disable byte-range windowing on the GPU, set"

Author: Done

[`spark.rapids.sql.window.range.byte.enabled`](configs.md#sql.window.range.byte.enabled) to true.

> We also provide the similar configurations:

Reviewer: ```suggestion
We also provide configurations for other integral range types:
```

Author: Done

[`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled),
[`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled),

> [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled).

Reviewer: Suggest using bullets here rather than a comma-separated list

Author: Done

the range boundary calculated for a value has overflow, CPU and GPU will get different results.

For example, consider the following dataset:

``` console
+------+---------+
| id   | dollars |
+------+---------+
|    1 |    NULL |
|    1 |      13 |
|    1 |      14 |
|    1 |      15 |
|    1 |      15 |
|    1 |      17 |
|    1 |      18 |
|    1 |      52 |
|    1 |      53 |
|    1 |      61 |
|    1 |      65 |
|    1 |      72 |
|    1 |      73 |
|    1 |      75 |
|    1 |      78 |
|    1 |      84 |
|    1 |      85 |
|    1 |      86 |
|    1 |      92 |
|    1 |      98 |
+------+---------+
```

After executing the SQL statement:

``` sql
SELECT
 COUNT(dollars) over
    (PARTITION BY id
    ORDER BY CAST (dollars AS Byte) ASC
    RANGE BETWEEN 127 PRECEDING AND 127 FOLLOWING)
FROM table
```

The results will differ between the CPU and GPU due to overflow handling.

``` console
CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0])
GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
```

> To disable byte-range windowing on the GPU, set

Reviewer 2: I think you mean enable not disable, either that or set the value to false.

Author: Thx for the finding. Done

[`spark.rapids.sql.window.range.byte.enabled`](configs.md#sql.window.range.byte.enabled) to true.

We also provide configurations for other integral range types:

- [`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled)
- [`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled)
- [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled)

> 

Reviewer 2: Can we have something explaining the defaults. It is not clear form this that ints and longs are enabled, but shorts and bytes are disabled, or why.

Author: Done.

GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
```

To disable byte-range windowing on the GPU, set
[`spark.rapids.sql.window.range.byte.enabled`](configs.md#sql.window.range.byte.enabled) to true.

We also provide configurations for other integral range types:

> - [`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled),
> - [`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled),
> - [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled).

Reviewer: ```suggestion
- [`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled)
- [`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled)
- [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled)
```
